### ISSUE TITLE
Bug in SlevomatCodingStandard.Variables.UnusedVariable when iterating oder self()

### ISSUE THREAD

**Issue by paslandau** [slevomat/coding-standard#578]
Noticed an edgecase where `SlevomatCodingStandard.Variables.UnusedVariable` is reporting a false positive when `ignoreUnusedValuesWhenOnlyKeysAreUsedInForeach` is set to true. This only happens when iterating over `new self()` (see code example).

Possibly related https://github.com/slevomat/coding-standard/issues/536

Ruleset
===
```xml
<?xml version="1.0"?>
<ruleset name="Error">
    <!-- relative path from PHPCS source location -->
    <config name="installed_paths" value="../../slevomat/coding-standard"/>
    <rule ref="SlevomatCodingStandard.Variables.UnusedVariable">
        <properties>
            <property name="ignoreUnusedValuesWhenOnlyKeysAreUsedInForeach" value="true"/>
        </properties>
    </rule>
</ruleset>
```

Code
==
```php
<?php

class Foo {
    function foo(){
        foreach (new self() as $key => $value) {
            echo $key;
        }
    }
}
```

Output
===
```
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 5 | ERROR | Unused variable $value.
----------------------------------------------------------------------
```

Output (long)
===
```
root@cbae699fa362:/var/www/current# vendor/bin/phpcs test2.php --standard=qa/phpcs/error.xml -vvv
Processing ruleset /var/www/current/qa/phpcs/error.xml
        => set config value installed_paths: ../../slevomat/coding-standard
        Processing rule "SlevomatCodingStandard.Variables.UnusedVariable"
                => /var/www/current/vendor/slevomat/coding-standard/SlevomatCodingStandard/Sniffs/Variables/UnusedVariableSniff.php
                => property "ignoreUnusedValuesWhenOnlyKeysAreUsedInForeach" set to "true"
=> Ruleset processing complete; included 1 sniffs and excluded 0
Registered SlevomatCodingStandard\Sniffs\Variables\UnusedVariableSniff
Creating file list... DONE (1 files in queue)
Changing into directory /var/www/current
Processing test2.php
        *** START PHP TOKENIZING ***
        Process token [0]: T_OPEN_TAG => <?php\n
        Process token [1]: T_WHITESPACE => \n
        Process token [2]: T_CLASS => class
        Process token [3]: T_WHITESPACE => ·
        Process token [4]: T_STRING => Foo
        Process token [5]: T_WHITESPACE => ·
        Process token  6 : T_OPEN_CURLY_BRACKET => {
        Process token [7]: T_WHITESPACE => \n····
        Process token [8]: T_FUNCTION => function
        Process token [9]: T_WHITESPACE => ·
        Process token [10]: T_STRING => foo
        Process token  11 : T_OPEN_PARENTHESIS => (
        Process token  12 : T_CLOSE_PARENTHESIS => )
        Process token  13 : T_OPEN_CURLY_BRACKET => {
        Process token [14]: T_WHITESPACE => \n········
        Process token [15]: T_FOREACH => foreach
        Process token [16]: T_WHITESPACE => ·
        Process token  17 : T_OPEN_PARENTHESIS => (
        Process token [18]: T_NEW => new
        Process token [19]: T_WHITESPACE => ·
        Process token [20]: T_STRING => self
        Process token  21 : T_OPEN_PARENTHESIS => (
        Process token  22 : T_CLOSE_PARENTHESIS => )
        Process token [23]: T_WHITESPACE => ·
        Process token [24]: T_AS => as
        Process token [25]: T_WHITESPACE => ·
        Process token [26]: T_VARIABLE => $key
        Process token [27]: T_WHITESPACE => ·
        Process token [28]: T_DOUBLE_ARROW => =>
        Process token [29]: T_WHITESPACE => ·
        Process token [30]: T_VARIABLE => $value
        Process token  31 : T_CLOSE_PARENTHESIS => )
        Process token [32]: T_WHITESPACE => ·
        Process token  33 : T_OPEN_CURLY_BRACKET => {
        Process token [34]: T_WHITESPACE => \n············
        Process token [35]: T_ECHO => echo
        Process token [36]: T_WHITESPACE => ·
        Process token [37]: T_VARIABLE => $key
        Process token  38 : T_SEMICOLON => ;
        Process token [39]: T_WHITESPACE => \n········
        Process token  40 : T_CLOSE_CURLY_BRACKET => }
        Process token [41]: T_WHITESPACE => \n····
        Process token  42 : T_CLOSE_CURLY_BRACKET => }
        Process token [43]: T_WHITESPACE => \n
        Process token  44 : T_CLOSE_CURLY_BRACKET => }
        Process token [45]: T_WHITESPACE => \n
        *** END PHP TOKENIZING ***
        *** START TOKEN MAP ***
        => Found curly bracket opener at 6
                => Found curly bracket opener at 14
                        => Found curly bracket opener at 35
                        => Found curly bracket closer at 44 for 35
                => Found curly bracket closer at 47 for 14
        => Found curly bracket closer at 49 for 6
        *** END TOKEN MAP ***
        *** START SCOPE MAP ***
        Start scope map at 2:T_CLASS => class
        => Begin scope map recursion at token 2 with depth 1
        Process token 3 on line 3 []: T_WHITESPACE => ·
        Process token 4 on line 3 []: T_STRING => Foo
        Process token 5 on line 3 []: T_WHITESPACE => ·
        Process token 6 on line 3 []: T_OPEN_CURLY_BRACKET => {
        => Found scope opener for 2:T_CLASS
        Process token 7 on line 3 [opener:6;]: T_WHITESPACE => \n
        Process token 8 on line 4 [opener:6;]: T_WHITESPACE => ····
        Process token 9 on line 4 [opener:6;]: T_FUNCTION => function
        * token is an opening condition *
        * searching for opener *
                => Begin scope map recursion at token 9 with depth 2
                Process token 10 on line 4 []: T_WHITESPACE => ·
                Process token 11 on line 4 []: T_STRING => foo
                Process token 12 on line 4 []: T_OPEN_PARENTHESIS => (
                Process token 13 on line 4 []: T_CLOSE_PARENTHESIS => )
                Process token 14 on line 4 []: T_OPEN_CURLY_BRACKET => {
                => Found scope opener for 9:T_FUNCTION
                Process token 15 on line 4 [opener:14;]: T_WHITESPACE => \n
                Process token 16 on line 5 [opener:14;]: T_WHITESPACE => ········
                Process token 17 on line 5 [opener:14;]: T_FOREACH => foreach
                * token is an opening condition *
                * searching for opener *
                        => Begin scope map recursion at token 17 with depth 3
                        Process token 18 on line 5 []: T_WHITESPACE => ·
                        Process token 19 on line 5 []: T_OPEN_PARENTHESIS => (
                        Process token 20 on line 5 []: T_NEW => new
                        Process token 21 on line 5 []: T_WHITESPACE => ·
                        Process token 22 on line 5 []: T_SELF => self
                        Process token 23 on line 5 []: T_OPEN_PARENTHESIS => (
                        Process token 24 on line 5 []: T_CLOSE_PARENTHESIS => )
                        Process token 25 on line 5 []: T_WHITESPACE => ·
                        Process token 26 on line 5 []: T_AS => as
                        Process token 27 on line 5 []: T_WHITESPACE => ·
                        Process token 28 on line 5 []: T_VARIABLE => $key
                        Process token 29 on line 5 []: T_WHITESPACE => ·
                        Process token 30 on line 5 []: T_DOUBLE_ARROW => =>
                        Process token 31 on line 5 []: T_WHITESPACE => ·
                        Process token 32 on line 5 []: T_VARIABLE => $value
                        Process token 33 on line 5 []: T_CLOSE_PARENTHESIS => )
                        Process token 34 on line 5 []: T_WHITESPACE => ·
                        Process token 35 on line 5 []: T_OPEN_CURLY_BRACKET => {
                        => Found scope opener for 17:T_FOREACH
                        Process token 36 on line 5 [opener:35;]: T_WHITESPACE => \n
                        Process token 37 on line 6 [opener:35;]: T_WHITESPACE => ············
                        Process token 38 on line 6 [opener:35;]: T_ECHO => echo
                        Process token 39 on line 6 [opener:35;]: T_WHITESPACE => ·
                        Process token 40 on line 6 [opener:35;]: T_VARIABLE => $key
                        Process token 41 on line 6 [opener:35;]: T_SEMICOLON => ;
                        Process token 42 on line 6 [opener:35;]: T_WHITESPACE => \n
                        Process token 43 on line 7 [opener:35;]: T_WHITESPACE => ········
                        Process token 44 on line 7 [opener:35;]: T_CLOSE_CURLY_BRACKET => }
                        => Found scope closer (44:T_CLOSE_CURLY_BRACKET) for 17:T_FOREACH
                Process token 45 on line 7 [opener:14;]: T_WHITESPACE => \n
                Process token 46 on line 8 [opener:14;]: T_WHITESPACE => ····
                Process token 47 on line 8 [opener:14;]: T_CLOSE_CURLY_BRACKET => }
                => Found scope closer (47:T_CLOSE_CURLY_BRACKET) for 9:T_FUNCTION
        Process token 48 on line 8 [opener:6;]: T_WHITESPACE => \n
        Process token 49 on line 9 [opener:6;]: T_CLOSE_CURLY_BRACKET => }
        => Found scope closer (49:T_CLOSE_CURLY_BRACKET) for 2:T_CLASS
        *** END SCOPE MAP ***
        *** START LEVEL MAP ***
        Process token 0 on line 1 [col:1;len:5;lvl:0;]: T_OPEN_TAG => <?php\n
        Process token 1 on line 2 [col:1;len:0;lvl:0;]: T_WHITESPACE => \n
        Process token 2 on line 3 [col:1;len:5;lvl:0;]: T_CLASS => class
        Process token 3 on line 3 [col:6;len:1;lvl:0;]: T_WHITESPACE => ·
        Process token 4 on line 3 [col:7;len:3;lvl:0;]: T_STRING => Foo
        Process token 5 on line 3 [col:10;len:1;lvl:0;]: T_WHITESPACE => ·
        Process token 6 on line 3 [col:11;len:1;lvl:0;]: T_OPEN_CURLY_BRACKET => {
        => Found scope opener for 2:T_CLASS
                * level increased *
                * token 2:T_CLASS added to conditions array *
                Process token 7 on line 3 [col:12;len:0;lvl:1;conds;T_CLASS;]: T_WHITESPACE => \n
                Process token 8 on line 4 [col:1;len:4;lvl:1;conds;T_CLASS;]: T_WHITESPACE => ····
                Process token 9 on line 4 [col:5;len:8;lvl:1;conds;T_CLASS;]: T_FUNCTION => function
                Process token 10 on line 4 [col:13;len:1;lvl:1;conds;T_CLASS;]: T_WHITESPACE => ·
                Process token 11 on line 4 [col:14;len:3;lvl:1;conds;T_CLASS;]: T_STRING => foo
                Process token 12 on line 4 [col:17;len:1;lvl:1;conds;T_CLASS;]: T_OPEN_PARENTHESIS => (
                Process token 13 on line 4 [col:18;len:1;lvl:1;conds;T_CLASS;]: T_CLOSE_PARENTHESIS => )
                Process token 14 on line 4 [col:19;len:1;lvl:1;conds;T_CLASS;]: T_OPEN_CURLY_BRACKET => {
                => Found scope opener for 9:T_FUNCTION
                        * level increased *
                        * token 9:T_FUNCTION added to conditions array *
                        Process token 15 on line 4 [col:20;len:0;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_WHITESPACE => \n
                        Process token 16 on line 5 [col:1;len:8;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_WHITESPACE => ········
                        Process token 17 on line 5 [col:9;len:7;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_FOREACH => foreach
                        Process token 18 on line 5 [col:16;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_WHITESPACE => ·
                        Process token 19 on line 5 [col:17;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_OPEN_PARENTHESIS => (
                        Process token 20 on line 5 [col:18;len:3;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_NEW => new
                        Process token 21 on line 5 [col:21;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_WHITESPACE => ·
                        Process token 22 on line 5 [col:22;len:4;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_SELF => self
                        Process token 23 on line 5 [col:26;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_OPEN_PARENTHESIS => (
                        Process token 24 on line 5 [col:27;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_CLOSE_PARENTHESIS => )
                        Process token 25 on line 5 [col:28;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_WHITESPACE => ·
                        Process token 26 on line 5 [col:29;len:2;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_AS => as
                        Process token 27 on line 5 [col:31;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_WHITESPACE => ·
                        Process token 28 on line 5 [col:32;len:4;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_VARIABLE => $key
                        Process token 29 on line 5 [col:36;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_WHITESPACE => ·
                        Process token 30 on line 5 [col:37;len:2;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_DOUBLE_ARROW => =>
                        Process token 31 on line 5 [col:39;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_WHITESPACE => ·
                        Process token 32 on line 5 [col:40;len:6;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_VARIABLE => $value
                        Process token 33 on line 5 [col:46;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_CLOSE_PARENTHESIS => )
                        Process token 34 on line 5 [col:47;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_WHITESPACE => ·
                        Process token 35 on line 5 [col:48;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_OPEN_CURLY_BRACKET => {
                        => Found scope opener for 17:T_FOREACH
                                * level increased *
                                * token 17:T_FOREACH added to conditions array *
                                Process token 36 on line 5 [col:49;len:0;lvl:3;conds;T_CLASS,T_FUNCTION,T_FOREACH;]: T_WHITESPACE => \n
                                Process token 37 on line 6 [col:1;len:12;lvl:3;conds;T_CLASS,T_FUNCTION,T_FOREACH;]: T_WHITESPACE => ············
                                Process token 38 on line 6 [col:13;len:4;lvl:3;conds;T_CLASS,T_FUNCTION,T_FOREACH;]: T_ECHO => echo
                                Process token 39 on line 6 [col:17;len:1;lvl:3;conds;T_CLASS,T_FUNCTION,T_FOREACH;]: T_WHITESPACE => ·
                                Process token 40 on line 6 [col:18;len:4;lvl:3;conds;T_CLASS,T_FUNCTION,T_FOREACH;]: T_VARIABLE => $key
                                Process token 41 on line 6 [col:22;len:1;lvl:3;conds;T_CLASS,T_FUNCTION,T_FOREACH;]: T_SEMICOLON => ;
                                Process token 42 on line 6 [col:23;len:0;lvl:3;conds;T_CLASS,T_FUNCTION,T_FOREACH;]: T_WHITESPACE => \n
                                Process token 43 on line 7 [col:1;len:8;lvl:3;conds;T_CLASS,T_FUNCTION,T_FOREACH;]: T_WHITESPACE => ········
                                Process token 44 on line 7 [col:9;len:1;lvl:3;conds;T_CLASS,T_FUNCTION,T_FOREACH;]: T_CLOSE_CURLY_BRACKET => }
                                => Found scope closer for 35:T_OPEN_CURLY_BRACKET
                                * token T_FOREACH removed from conditions array *
                                * level decreased *
                        Process token 45 on line 7 [col:10;len:0;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_WHITESPACE => \n
                        Process token 46 on line 8 [col:1;len:4;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_WHITESPACE => ····
                        Process token 47 on line 8 [col:5;len:1;lvl:2;conds;T_CLASS,T_FUNCTION;]: T_CLOSE_CURLY_BRACKET => }
                        => Found scope closer for 14:T_OPEN_CURLY_BRACKET
                        * token T_FUNCTION removed from conditions array *
                        * level decreased *
                Process token 48 on line 8 [col:6;len:0;lvl:1;conds;T_CLASS;]: T_WHITESPACE => \n
                Process token 49 on line 9 [col:1;len:1;lvl:1;conds;T_CLASS;]: T_CLOSE_CURLY_BRACKET => }
                => Found scope closer for 6:T_OPEN_CURLY_BRACKET
                * token T_CLASS removed from conditions array *
                * level decreased *
        Process token 50 on line 9 [col:2;len:0;lvl:0;]: T_WHITESPACE => \n
        *** END LEVEL MAP ***
        *** START ADDITIONAL PHP PROCESSING ***
        *** END ADDITIONAL PHP PROCESSING ***
[PHP => 51 tokens in 9 lines]...
        *** START TOKEN PROCESSING ***
                Process token 0: T_OPEN_TAG => <?php\n
                Process token 1: T_WHITESPACE => \n
                Process token 2: T_CLASS => class
                Process token 3: T_WHITESPACE => ·
                Process token 4: T_STRING => Foo
                Process token 5: T_WHITESPACE => ·
                Process token 6: T_OPEN_CURLY_BRACKET => {
                Process token 7: T_WHITESPACE => \n
                Process token 8: T_WHITESPACE => ····
                Process token 9: T_FUNCTION => function
                Process token 10: T_WHITESPACE => ·
                Process token 11: T_STRING => foo
                Process token 12: T_OPEN_PARENTHESIS => (
                Process token 13: T_CLOSE_PARENTHESIS => )
                Process token 14: T_OPEN_CURLY_BRACKET => {
                Process token 15: T_WHITESPACE => \n
                Process token 16: T_WHITESPACE => ········
                Process token 17: T_FOREACH => foreach
                Process token 18: T_WHITESPACE => ·
                Process token 19: T_OPEN_PARENTHESIS => (
                Process token 20: T_NEW => new
                Process token 21: T_WHITESPACE => ·
                Process token 22: T_SELF => self
                Process token 23: T_OPEN_PARENTHESIS => (
                Process token 24: T_CLOSE_PARENTHESIS => )
                Process token 25: T_WHITESPACE => ·
                Process token 26: T_AS => as
                Process token 27: T_WHITESPACE => ·
                Process token 28: T_VARIABLE => $key
                        Processing SlevomatCodingStandard\Sniffs\Variables\UnusedVariableSniff... DONE in 0.0249 seconds
                Process token 29: T_WHITESPACE => ·
                Process token 30: T_DOUBLE_ARROW => =>
                Process token 31: T_WHITESPACE => ·
                Process token 32: T_VARIABLE => $value
                        Processing SlevomatCodingStandard\Sniffs\Variables\UnusedVariableSniff... DONE in 0.0002 seconds
                Process token 33: T_CLOSE_PARENTHESIS => )
                Process token 34: T_WHITESPACE => ·
                Process token 35: T_OPEN_CURLY_BRACKET => {
                Process token 36: T_WHITESPACE => \n
                Process token 37: T_WHITESPACE => ············
                Process token 38: T_ECHO => echo
                Process token 39: T_WHITESPACE => ·
                Process token 40: T_VARIABLE => $key
                        Processing SlevomatCodingStandard\Sniffs\Variables\UnusedVariableSniff... DONE in 0 seconds
                Process token 41: T_SEMICOLON => ;
                Process token 42: T_WHITESPACE => \n
                Process token 43: T_WHITESPACE => ········
                Process token 44: T_CLOSE_CURLY_BRACKET => }
                Process token 45: T_WHITESPACE => \n
                Process token 46: T_WHITESPACE => ····
                Process token 47: T_CLOSE_CURLY_BRACKET => }
                Process token 48: T_WHITESPACE => \n
                Process token 49: T_CLOSE_CURLY_BRACKET => }
                Process token 50: T_WHITESPACE => \n
        *** END TOKEN PROCESSING ***
        *** START SNIFF PROCESSING REPORT ***
        SlevomatCodingStandard\Sniffs\Variables\UnusedVariableSniff: 0.0251 secs
        *** END SNIFF PROCESSING REPORT ***
DONE in 68ms (1 errors, 0 warnings)

FILE: /var/www/current/test2.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 5 | ERROR | Unused variable $value.
----------------------------------------------------------------------

Time: 260ms; Memory: 4MB

root@cbae699fa362:/var/www/current# vendor/bin/phpcs test2.php --standard=qa/phpcs/error.xml -v
Registering sniffs in the Error standard... DONE (1 sniffs registered)
Creating file list... DONE (1 files in queue)
Changing into directory /var/www/current
Processing test2.php [PHP => 51 tokens in 9 lines]... DONE in 75ms (1 errors, 0 warnings)

FILE: /var/www/current/test2.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 5 | ERROR | Unused variable $value.
----------------------------------------------------------------------

Time: 254ms; Memory: 4MB
```

**Comment by kukulich** [comment]
Thank you for your report.

I think it should be fixed in https://github.com/slevomat/coding-standard/commit/7b401c237697f0d3a8e447627fa52725f9c5ecd6
